jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java

```diff
@@ -32,112 +32,132 @@
+      extractionPathMap.put(LayerType.DEPENDENCIES, extractionPath);
       return this;
     }
 
-    public Builder setSnapshotDependencyFiles(List<Path> snapshotDependencyFiles) {
+    public Builder setSnapshotDependencyFiles(
+        List<Path> snapshotDependencyFiles, String extractionPath) {
       layerFilesMap.put(LayerType.SNAPSHOT_DEPENDENCIES, snapshotDependencyFiles);
+      extractionPathMap.put(LayerType.SNAPSHOT_DEPENDENCIES, extractionPath);
       return this;
     }
 
-    public Builder setResourceFiles(List<Path> resourceFiles) {
+    public Builder setResourceFiles(List<Path> resourceFiles, String extractionPath) {
       layerFilesMap.put(LayerType.RESOURCES, resourceFiles);
+      extractionPathMap.put(LayerType.RESOURCES, extractionPath);
       return this;
     }
 
-    public Builder setClassFiles(List<Path> classFiles) {
+    public Builder setClassFiles(List<Path> classFiles, String extractionPath) {
       layerFilesMap.put(LayerType.CLASSES, classFiles);
+      extractionPathMap.put(LayerType.CLASSES, extractionPath);
       return this;
     }
 
-    public Builder setExtraFiles(List<Path> extraFiles) {
+    public Builder setExtraFiles(List<Path> extraFiles, String extractionPath) {
       layerFilesMap.put(LayerType.EXTRA_FILES, extraFiles);
-      return this;
-    }
-
-    // TODO: remove this and put files in WAR into the relevant layers (i.e., dependencies, snapshot
-    // dependencies, resources, and classes layers).
-    public Builder setExplodedWarFiles(List<Path> explodedWarFiles) {
-      layerFilesMap.put(LayerType.EXPLODED_WAR, explodedWarFiles);
+      extractionPathMap.put(LayerType.EXTRA_FILES, extractionPath);
       return this;
     }
 
     public JavaLayerConfigurations build() throws IOException {
       ImmutableMap.Builder<LayerType, LayerConfiguration> layerConfigurationsMap =
           ImmutableMap.builderWithExpectedSize(LayerType.values().length);
       for (LayerType layerType : LayerType.values()) {
+        String extractionPath = Preconditions.checkNotNull(extractionPathMap.get(layerType));
+        // Windows filenames cannot have "/", so this also blocks Windows-style path.
+        Preconditions.checkState(
+            extractionPath.startsWith("/"),
+            "extractionPath should be an absolute path in Unix-style: " + extractionPath);
+
         LayerConfiguration.Builder layerConfigurationBuilder =
             LayerConfiguration.builder().setName(layerType.getName());
 
         // Adds all the layer files recursively.
         List<Path> layerFiles = Preconditions.checkNotNull(layerFilesMap.get(layerType));
         for (Path layerFile : layerFiles) {
-          layerConfigurationBuilder.addEntryRecursive(
-              layerFile, layerType.getExtractionPath().resolve(layerFile.getFileName()));
+          final Path pathInContainer;
+          Path subPath = Paths.get(extractionPath);
+          if (layerType.equals(LayerType.RESOURCES)
```

Comment:
Urgh... I guess #984 isn't enough. I'll fix #1007, and after that, I think this class will become truly generic and there won't be any special code to handle the WAR case.